In [1]:
# Import Packages

# Array and Dataframe processing:
import pandas as pd
import numpy as np

# Visualization:
import seaborn as sns
sns.set_style("darkgrid")
import matplotlib.pyplot as plt
%matplotlib inline
import plotly.graph_objects as go
import plotly.express as px
import plotly.figure_factory as ff

import pandas as pd
import mysql.connector 

In [2]:
# # Model Implementation 
# from sklearn.ensemble import RandomForestClassifier
# from sklearn import svm
# from sklearn.svm import SVC
# from sklearn.neural_network import MLPClassifier
# # from sklearn.linear_model import SGDCClassifier

# #  Metrics 
# from sklearn.metrics import confusion_matrix, classification_report
# from sklearn.metrics import accuracy_score

# # For scaling and preprocessing:  StandardScaler, MinMaxScaler and MaxAbsScaler.
# from sklearn import preprocessing
# from sklearn.preprocessing import StandardScaler, LabelEncoder

# # For Model Training and Testing:
# from sklearn.model_selection import train_test_split

In [3]:
df = pd.read_csv('../Data/HR_Data.csv')
df

EmpID           Employee_Name  MarriedID  MaritalStatusID  GenderID  \
0    1103024456              Brown, Mia          1                1         0   
1    1106026572    LaRotonda, William            0                2         1   
2    1302053333        Steans, Tyrone            0                0         1   
3    1211050782         Howard, Estelle          1                1         0   
4    1307059817             Singh, Nan           0                0         0   
..          ...                     ...        ...              ...       ...   
305  1411071312          Daniele, Ann            1                1         0   
306  1108028108          Lajiri,  Jyoti          1                1         1   
307   904013591  Semizoglou, Jeremiah            1                1         1   
308  1308060959              South, Joe          0                0         1   
309  1301052347         Warfield, Sarah          0                4         0   

     EmpStatusID  DeptID  PerfScoreID  FromDiversityJobFairID  PayRate  ...  \
0              1       1            3                       1    28.50  ...   
1              1       1            3                       0    23.00  ...   
2              1       1            3                       0    29.00  ...   
3              1       1            3                       0    21.50  ...   
4              1       1            3                       0    16.56  ...   
..           ...     ...          ...                     ...      ...  ...   
305            3       3            3                       0    54.10  ...   
306            3       3            3                       0    56.20  ...   
307            2       3            3                       0    53.80  ...   
308            1       3            3                       0    53.00  ...   
309            1       3            3                       0    55.20  ...   

        Department         ManagerName ManagerID       RecruitmentSource  \
0    Admin Offices  Brandon R. LeBlanc       1.0      Diversity Job Fair   
1    Admin Offices  Brandon R. LeBlanc       1.0      Website Banner Ads   
2    Admin Offices  Brandon R. LeBlanc       1.0         Internet Search   
3    Admin Offices  Brandon R. LeBlanc       1.0  Pay Per Click - Google   
4    Admin Offices  Brandon R. LeBlanc       1.0      Website Banner Ads   
..             ...                 ...       ...                     ...   
305          IT/IS        Peter Monroe       7.0         Vendor Referral   
306          IT/IS        Peter Monroe       7.0       Employee Referral   
307          IT/IS        Peter Monroe       7.0    On-campus Recruiting   
308          IT/IS        Peter Monroe       7.0       Employee Referral   
309          IT/IS        Peter Monroe       7.0       Employee Referral   

     PerformanceScore EngagementSurvey EmpSatisfaction SpecialProjectsCount  \
0         Fully Meets             2.04               2                    6   
1         Fully Meets             5.00               4                    4   
2         Fully Meets             3.90               5                    5   
3         Fully Meets             3.24               3                    4   
4         Fully Meets             5.00               3                    5   
..                ...              ...             ...                  ...   
305       Fully Meets             1.55               5                    8   
306       Fully Meets             4.46               5                    6   
307       Fully Meets             3.27               4                    5   
308       Fully Meets             5.00               3                    4   
309       Fully Meets             1.06               4                    8   

    LastPerformanceReview_Date DaysLateLast30  
0                      1/15/19            0.0  
1                      1/17/19            0.0  
2                      1/18/19            0.0  
3                          NaN    

In [4]:
dic = pd.read_excel('../Data/data_dictionary.xlsx')
dic

Feature  \
0               Employee Name   
1                       EmpID   
2                   MarriedID   
3             MaritalStatusID   
4                 EmpStatusID   
5                      DeptID   
6                 PerfScoreID   
7      FromDiversityJobFairID   
8                     PayRate   
9                       Termd   
10                 PositionID   
11                   Position   
12                      State   
13                        Zip   
14                        DOB   
15                        Sex   
16                MaritalDesc   
17                CitizenDesc   
18             HispanicLatino   
19                   RaceDesc   
20                 DateofHire   
21          DateofTermination   
22                 TermReason   
23           EmploymentStatus   
24                 Department   
25                ManagerName   
26                  ManagerID   
27          RecruitmentSource   
28           PerformanceScore   
29           EngagementSurvey   
30            EmpSatisfaction   
31       SpecialProjectsCount   
32  LastPerformanceReviewDate   
33             DaysLateLast30   

                                          Description DataType  
0                                Employee’s full name     Text  
1              Employee ID is unique to each employee     Text  
2        Is the person married (1 or 0 for yes or no)   Binary  
3   Marital status code that matches the text fiel...  Integer  
4   Employment status code that matches text field...  Integer  
5   Department ID code that matches the department...  Integer  
6   Performance Score code that matches the employ...  Integer  
7   Was the employee sourced from the Diversity jo...   Binary  
8   The person’s hourly pay rate. All salaries are...    Float  
9          Has this employee been terminated - 1 or 0   Binary  
10        An integer indicating the person’s position  Integer  
11  The text name/title of the position the person...     Text  
12                 The state that the person lives in     Text  
13                      The zip code for the employee     Text  
14                     Date of Birth for the employee     Date  
15                                       Sex - M or F     Text  
16  The marital status of the person (divorced, si...     Text  
17  Label for whether the person is a Citizen or E...     Text  
18  Yes or No field for whether the employee is Hi...     Text  
19  Description/text of the race the person identi...     Text  
20                          Date the person was hired     Date  
21  Date the person was terminated, only populated...     Date  
22  A text reason / description for why the person...     Text  
23  A description/category of the person’s employm...     Text  
24    Name of the department that the person works in     Text  
25         The name of the person’s immediate manager     Text  
26              A unique identifier for each manager.  Integer  
27  The name of the recruitment source where the e...     Text  
28  Performance Score text/category (Fully Meets, ...     Text  
29  Results from the last engagement survey, manag...    Float  
30  A basic satisfaction score between 1 and 5, as...  Integer  
31  The number of special projects that the employ...  Integer  
32  The most recent date of the person’s last perf...     Date  
33  The number of times that the employee was late...  Integer

In [5]:
print(f'Dataframe Shape: {df.shape}')
df.describe()

Dataframe Shape: (310, 35)


EmpID   MarriedID  MaritalStatusID    GenderID  EmpStatusID  \
count  3.100000e+02  310.000000       310.000000  310.000000   310.000000   
mean   1.199745e+09    0.396774         0.809677    0.429032     2.396774   
std    1.829600e+08    0.490019         0.944702    0.495738     1.795533   
min    6.020003e+08    0.000000         0.000000    0.000000     1.000000   
25%    1.101024e+09    0.000000         0.000000    0.000000     1.000000   
50%    1.203032e+09    0.000000         1.000000    0.000000     1.000000   
75%    1.378814e+09    1.000000         1.000000    1.000000     5.000000   
max    1.988300e+09    1.000000         4.000000    1.000000     5.000000   

           DeptID  PerfScoreID  FromDiversityJobFairID     PayRate  \
count  310.000000   310.000000              310.000000  310.000000   
mean     4.606452     2.983871                0.093548   31.284806   
std      1.082349     0.577124                0.291670   15.383615   
min      1.000000     1.000000                0.000000   14.000000   
25%      5.000000     3.000000                0.000000   20.000000   
50%      5.000000     3.000000                0.000000   24.000000   
75%      5.000000     3.000000                0.000000   45.315000   
max      6.000000     4.000000                1.000000   80.000000   

            Termd  PositionID           Zip   ManagerID  EngagementSurvey  \
count  310.000000  310.000000    310.000000  302.000000        310.000000   
mean     0.332258   16.838710   6569.732258   14.579470          3.332097   
std      0.471785    6.232272  16933.864054    8.090352          1.290590   
min      0.000000    1.000000   1013.000000    1.000000          1.030000   
25%      0.000000   18.000000   1901.250000   10.000000          2.082500   
50%      0.000000   19.000000   2132.000000   15.000000          3.470000   
75%      1.000000   20.000000   2357.000000   19.000000          4.520000   
max      1.000000   30.000000  98052.000000   39.000000          5.000000   

       EmpSatisfaction  SpecialProjectsCount  DaysLateLast30  
count       310.000000            310.000000           207.0  
mean          3.890323              1.209677             0.0  
std           0.910690              2.352009             0.0  
min           1.000000              0.000000             0.0  
25%           3.000000              0.000000             0.0  
50%           4.000000              0.000000             0.0  
75%           5.000000              0.000000             0.0  
max           5.000000              8.000000             0.0

In [6]:
df.isna().sum()

EmpID                           0
Employee_Name                   0
MarriedID                       0
MaritalStatusID                 0
GenderID                        0
EmpStatusID                     0
DeptID                          0
PerfScoreID                     0
FromDiversityJobFairID          0
PayRate                         0
Termd                           0
PositionID                      0
Position                        0
State                           0
Zip                             0
DOB                             0
Sex                             0
MaritalDesc                     0
CitizenDesc                     0
HispanicLatino                  0
RaceDesc                        0
DateofHire                      0
DateofTermination             207
TermReason                      1
EmploymentStatus                0
Department                      0
ManagerName                     0
ManagerID                       8
RecruitmentSource               0
PerformanceSco

There are three columns with null values:  Date of termination, Last Performance Review Date and Days late within the last thirty days.  Additionally, there is one column with a sole value missing, Termination reason. There are equal numbers of ommitted entries for last performance review and days late within the last 30 days.  These appear to be employees wh have been terminated; managers should have still input these fieldsbut neglected to do so.  Date of Termination was left blank for active employees as obviously they haven't been terminated yet.  Consequently, the null values for each column must be treated seperately.

In [7]:
df

EmpID           Employee_Name  MarriedID  MaritalStatusID  GenderID  \
0    1103024456              Brown, Mia          1                1         0   
1    1106026572    LaRotonda, William            0                2         1   
2    1302053333        Steans, Tyrone            0                0         1   
3    1211050782         Howard, Estelle          1                1         0   
4    1307059817             Singh, Nan           0                0         0   
..          ...                     ...        ...              ...       ...   
305  1411071312          Daniele, Ann            1                1         0   
306  1108028108          Lajiri,  Jyoti          1                1         1   
307   904013591  Semizoglou, Jeremiah            1                1         1   
308  1308060959              South, Joe          0                0         1   
309  1301052347         Warfield, Sarah          0                4         0   

     EmpStatusID  DeptID  PerfScoreID  FromDiversityJobFairID  PayRate  ...  \
0              1       1            3                       1    28.50  ...   
1              1       1            3                       0    23.00  ...   
2              1       1            3                       0    29.00  ...   
3              1       1            3                       0    21.50  ...   
4              1       1            3                       0    16.56  ...   
..           ...     ...          ...                     ...      ...  ...   
305            3       3            3                       0    54.10  ...   
306            3       3            3                       0    56.20  ...   
307            2       3            3                       0    53.80  ...   
308            1       3            3                       0    53.00  ...   
309            1       3            3                       0    55.20  ...   

        Department         ManagerName ManagerID       RecruitmentSource  \
0    Admin Offices  Brandon R. LeBlanc       1.0      Diversity Job Fair   
1    Admin Offices  Brandon R. LeBlanc       1.0      Website Banner Ads   
2    Admin Offices  Brandon R. LeBlanc       1.0         Internet Search   
3    Admin Offices  Brandon R. LeBlanc       1.0  Pay Per Click - Google   
4    Admin Offices  Brandon R. LeBlanc       1.0      Website Banner Ads   
..             ...                 ...       ...                     ...   
305          IT/IS        Peter Monroe       7.0         Vendor Referral   
306          IT/IS        Peter Monroe       7.0       Employee Referral   
307          IT/IS        Peter Monroe       7.0    On-campus Recruiting   
308          IT/IS        Peter Monroe       7.0       Employee Referral   
309          IT/IS        Peter Monroe       7.0       Employee Referral   

     PerformanceScore EngagementSurvey EmpSatisfaction SpecialProjectsCount  \
0         Fully Meets             2.04               2                    6   
1         Fully Meets             5.00               4                    4   
2         Fully Meets             3.90               5                    5   
3         Fully Meets             3.24               3                    4   
4         Fully Meets             5.00               3                    5   
..                ...              ...             ...                  ...   
305       Fully Meets             1.55               5                    8   
306       Fully Meets             4.46               5                    6   
307       Fully Meets             3.27               4                    5   
308       Fully Meets             5.00               3                    4   
309       Fully Meets             1.06               4                    8   

    LastPerformanceReview_Date DaysLateLast30  
0                      1/15/19            0.0  
1                      1/17/19            0.0  
2                      1/18/19            0.0  
3                          NaN    

In [8]:
df["TermReason"] = df["TermReason"].fillna("None")
df.isna().sum()

EmpID                           0
Employee_Name                   0
MarriedID                       0
MaritalStatusID                 0
GenderID                        0
EmpStatusID                     0
DeptID                          0
PerfScoreID                     0
FromDiversityJobFairID          0
PayRate                         0
Termd                           0
PositionID                      0
Position                        0
State                           0
Zip                             0
DOB                             0
Sex                             0
MaritalDesc                     0
CitizenDesc                     0
HispanicLatino                  0
RaceDesc                        0
DateofHire                      0
DateofTermination             207
TermReason                      0
EmploymentStatus                0
Department                      0
ManagerName                     0
ManagerID                       8
RecruitmentSource               0
PerformanceSco

In [9]:
df["DateofTermination"] = df["DateofTermination"].fillna("None")
df.isna().sum()

EmpID                           0
Employee_Name                   0
MarriedID                       0
MaritalStatusID                 0
GenderID                        0
EmpStatusID                     0
DeptID                          0
PerfScoreID                     0
FromDiversityJobFairID          0
PayRate                         0
Termd                           0
PositionID                      0
Position                        0
State                           0
Zip                             0
DOB                             0
Sex                             0
MaritalDesc                     0
CitizenDesc                     0
HispanicLatino                  0
RaceDesc                        0
DateofHire                      0
DateofTermination               0
TermReason                      0
EmploymentStatus                0
Department                      0
ManagerName                     0
ManagerID                       8
RecruitmentSource               0
PerformanceSco

In [10]:
df.columns

Index(['EmpID', 'Employee_Name', 'MarriedID', 'MaritalStatusID', 'GenderID',
       'EmpStatusID', 'DeptID', 'PerfScoreID', 'FromDiversityJobFairID',
       'PayRate', 'Termd', 'PositionID', 'Position', 'State', 'Zip', 'DOB',
       'Sex', 'MaritalDesc', 'CitizenDesc', 'HispanicLatino', 'RaceDesc',
       'DateofHire', 'DateofTermination', 'TermReason', 'EmploymentStatus',
       'Department', 'ManagerName', 'ManagerID', 'RecruitmentSource',
       'PerformanceScore', 'EngagementSurvey', 'EmpSatisfaction',
       'SpecialProjectsCount', 'LastPerformanceReview_Date', 'DaysLateLast30'],
      dtype='object')

### Initial Analysis

#### Diversity

##### Average Performance Score by Gender


In [41]:
# creating dataframe to analyse Mean Performance Score by Sex
scoring_mean = pd.DataFrame(df.groupby(['Sex'])['PerfScoreID'].mean())
scoring_mean = scoring_mean.reset_index()
scoring_mean =  scoring_mean.sort_values(by = 'PerfScoreID', ascending = True )
scoring_mean

Sex  PerfScoreID
1  M      2.947368
0   F     3.011299

In [42]:
# Using Plotly to Visualize Mean Performance Score by Manager ID:
fig = px.bar(scoring_mean, x= 'Sex', y = 'PerfScoreID',  color = 'PerfScoreID', 
             title='Mean Performance Score by Sex')
fig.show()

Women have a higher average performance score.

In [43]:
gender_count = pd.DataFrame(df.groupby(['Sex'])['PerfScoreID'].count())
gender_count = gender_count.reset_index()
gender_count =  gender_count.sort_values(by = 'PerfScoreID', ascending = True )
gender_count

Sex  PerfScoreID
1  M           133
0   F          177

In [44]:
# Using Plotly to Visualize Mean Performance Score by Manager ID:
fig = px.bar(gender_count, x= 'Sex', y = 'PerfScoreID',  color = 'Sex', 
             title='Employee Count by Gender')
fig.show()

Not only are mean performance scores higher for women, but women constitute a majority of employees.

In [13]:
# creating dataframe to analyse Mean Performance Score by Employee Satisfaction
performance_race = pd.DataFrame(df.groupby(['RaceDesc'])['PerfScoreID'].mean())
performance_race = performance_race.reset_index()
performance_race =  performance_race.sort_values(by = 'PerfScoreID', ascending = True )
performance_race

RaceDesc  PerfScoreID
3                          Hispanic     2.750000
2         Black or African American     2.912281
4                 Two or more races     2.944444
1                             Asian     3.000000
5                             White     3.000000
0  American Indian or Alaska Native     3.500000

In [15]:
# Using Plotly to Visualize Mean Performance Score by Manager ID:
fig = px.bar(performance_race, x= 'RaceDesc', y = 'PerfScoreID',  color = 'RaceDesc', 
             title='Mean Performance Score by Racial Description')
fig.show()

Surprisingly, Perfornamce scores for native americans are signifantly higher that any other racial category.  This is especially pecular because the mean performance score range for the rest of the racial groups is .25 whereas the mean performance score for Native americans is .5 higher than the next group, white employees.  There doesn't appear to be a reason for this except for the presence of an outlier. So I look at counts.

In [29]:
# creating dataframe to analyse Mean Performance Score by Employee Satisfacticon
count_performance_race = pd.DataFrame(df.groupby(['RaceDesc'])['PerfScoreID'].count())
count_performance_race = count_performance_race.reset_index()
count_performance_race =  count_performance_race.sort_values(by = 'PerfScoreID', ascending = True )

In [40]:
# Using Plotly to Visualize Mean Performance Score by Manager ID:
fig = px.bar(count_performance_race, x= 'RaceDesc', y = 'PerfScoreID',  color = 'RaceDesc', 
             title=' Performance Score by Racial Description')
fig.show()
print(count_performance_race)
print(df['EmpID'].sum)

                           RaceDesc  PerfScoreID
0  American Indian or Alaska Native            4
3                          Hispanic            4
4                 Two or more races           18
1                             Asian           34
2         Black or African American           57
5                             White          193
<bound method Series.sum of 0      1103024456
1      1106026572
2      1302053333
3      1211050782
4      1307059817
          ...    
305    1411071312
306    1108028108
307     904013591
308    1308060959
309    1301052347
Name: EmpID, Length: 310, dtype: int64>


When we review the counts, the story told is quite different.  Here we can see that there simply are very few native Americans who are employees, totalling four.  By contrast, there are 193 White employees and 57 black employees. Consequently, the mean score for native americans is affected to a far greater extent by an individual employees score than is a member of other groups.  

What might this mean?  It's hard to say.  My first thought was that this was a significant indicator ofdiscriminatory hiring practices because the company has employees in 28 states.  However, by looking at employee counts by state, we can see an overwhelming concentration of employees in one state.

In [36]:
# looking at employees by state
employee_by_state = pd.DataFrame(df.groupby(['State'])['EmpID'].count())
employee_by_state = employee_by_state.reset_index()
employee_by_state =  employee_by_state.sort_values(by = 'EmpID', ascending = True )

In [37]:
# Using Plotly to Visualize employees by state:
fig = px.bar(employee_by_state, x= 'State', y = 'EmpID',  color = 'State', 
             title=' Number of Employees by State')
fig.show()
print(employee_by_state)

   State  EmpID
0     AL      1
25    VA      1
24    UT      1
22    TN      1
21    RI      1
20    PA      1
19    OR      1
18    OH      1
17    NY      1
16    NV      1
15    NH      1
14    ND      1
13    NC      1
11    ME      1
9     KY      1
8     IN      1
7     ID      1
6     GA      1
5     FL      1
3     CO      1
2     CA      1
1     AZ      1
12    MT      1
27    WA      1
26    VT      2
23    TX      3
4     CT      6
10    MA    275


88% of employees are in one state- Massachusetts. Most of the rest of the 27 states with employees have no more than 1 employee, almost always a statewide sales representiative.  The state of Massachusetts has a population that is 80% white.   